active-rfcs/0000-vtu-api.md

```diff
@@ -0,0 +1,268 @@
+
+- **Breaking:** `sync` mode removed. All `wrapper` methods return `nextTick`, you must `await` them
+- **Breaking:** `find` is now split into `find` and `findComponent`.
+- **Breaking:** Removal of some wrapper properties and methods, as they induce bad testing habits or are obsolete. 
+- **Breaking:** `shallowMount` stubs defaults slots
+- **Breaking:** `setProps` only works for the mounted component. 
+
+**Note:** The API for VueTestUtils 1.x will stay the same and will support Vue 2.x. 
+
+# Motivation
+
+- Support Vue 3's Component APIs.
+- Provide a smaller and easier to understand API
+- Allow adding custom functionality via plugin system. [TODO]
+- Remove methods that are bloat or lead to bad testing habits.
+- Generally improve VTU Docs and Guides.
+
+# Detailed design
+
+- `createLocalVue` is removed. Vue now exposes `createApp`, which creates an isolated app instance. VTU does that under the hood.
+- Fully `async`, each method that involves a mutation returns `nextTick`. Methods like `setValue` and `trigger` can be awaited, ensuring the DOM is re-rendered before each assertion.
+- Rewritten completely in TypeScript, giving much improved type hints when writing tests.
+- `shallowMount` will stub default slots of stubbed components. There will be an opt-in configuration to enable rendering slots for stubbed components in a similar manner to VTU beta. There is a limitation that scoped slots will not be able to provide data in such cases.
+- Simple plugin system to allow extending VTU with your own methods. See [Plugins](#plugin-system)
+
+## API changes
+
+We will only list the changes and deprecations. Please check the temporary [documentation](https://vuejs.github.io/vue-test-utils-next-docs/) for a full API listing.
+
+### mountOptions
+
+#### props
+
+[Link](https://vuejs.github.io/vue-test-utils-next-docs/api/#props) - Renamed from `propsData` to match component `props` field.
+
+#### global
+
+[Link](https://vuejs.github.io/vue-test-utils-next-docs/api/#global) - The `global` namespace is used to pass configuration to the `createApp` instance. Things like global components, directives and so on.
+
+These settings can also be globally set via the exported `config` object - `config.global.mocks.$t = jest.fn()`.
+
+- **global.components** - register global components
+- **global.directives** - register a global directive
+- **global.mixins** - register a global mixin
+- **global.plugins** - install a plugin
+- **global.stubs** - see bellow
+- **global.mocks** - see bellow
+- **global.provide** - see bellow
+
+#### stubs 
+
+[Link](https://vuejs.github.io/vue-test-utils-next-docs/api/#global-stubs) - Moved to `global.stubs`. 
+
+- **New** - Stubs will no longer render the slots of a component. This can be enabled by a global flag `config.renderStubSlots = true`.
+
+#### mocks
+
+[Link](https://vuejs.github.io/vue-test-utils-next-docs/api/#global-mocks) - Moved to `global.mocks`
+
+#### provide
```

Comment:
While having everything affecting `global` namespace in `global` is perfect, thank you for this, I'm a bit concerned about `provide`. From what I understand We have two things in "global":

* things which are registered globally in real world application (components, directives, mixins, plugins)
* things which are VTU-specific but affect each component in the tree (stubs, mocks)

`provide` is just providing data for `provide/inject`, and for example providing `foo` in `global.provide` does not guarantee that my value will be always injected in component somewhere deep in the tree - it could be easily changed by `provide` in some ancestor

Maybe it's worth to keep this our of `global` object since its behavior is different?